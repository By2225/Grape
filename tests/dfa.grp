fun Int main() {

    Edge<String> aEdge;

    Node<Int> start = '0';

    String input = "bbbbbbbbbbb";
    Graph<Int, String> dfa = << start  -"b"-  '1'  -"b"-  '0'  -"b"-  '1' >>;
    Int token = 0;
    Int i = 0;
    Bool found = False;

    String character;
    List<Edge<String> > transitions;

    Node<Int> state = start;
    while( token < str_size(input)) { 			// String size

	character = get_char(input, token);		// Get char at token
        transitions = get_outgoing(state); 			// Get outgoing edges

	i = 0;
	while ((i < size(transitions)) and (found == False)) { 			// List size
	    aEdge = transitions[i]; 	// Get list item
	    if (str_equal(aEdge.val, character)) { 	// Get list string
		state = get_to(aEdge);			// Get edge destination
                found = True;
	    }
	    i = i + 1;
	}
        if ( found == False){
            prints("REJECT");
            return 0;
        }
        found = False;
	token = token + 1;
    }

    if (state.val == 1) {		// Get accepting
        prints("ACCEPT");
    } else {
        prints("REJECT");
    }

    return 0;
}
