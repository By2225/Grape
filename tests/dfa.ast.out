
fun Int main() {
Edge<String> aEdge;
Node<Int> start = '0';
String input = "bbbbbbbbbbb";
Graph<Node<Int>, Edge<String>> dfa = <<start -"b"- '1' -"b"- '0' -"b"- '1' >>;
Int token = 0;
Int i = 0;
Bool found = false;
String character;
List<Edge<String>> transitions;
Node<Int> state = start;
while (token < str_size(input)) {
character = get_char(token, input);
transitions = get_outgoing(state);
i = 0;
while (i < size(transitions) and found == false) {
aEdge = transitions[i];
if (str_equal(character, aEdge.val))
{
state = get_to(aEdge);
found = true;
}
i = i + 1;
}
if (found == false)
{
prints("REJECT");
return 0;
}
found = false;
token = token + 1;
}
if (state.val == 1)
{
prints("ACCEPT");
}
else
{
prints("REJECT");
}
return 0;
}
